src/curl_tracker_client.cpp

```diff
@@ -0,0 +1,435 @@
+
+			if (len % 18 != 0 || len < 0)
+			{
+				// Invalid IPv6 peer data - must be multiple of 18 bytes
+				// Skip invalid data rather than failing entire response
+				// This is more robust for partially corrupt responses
+			}
+			else if (len > 0)
+			{
+				// Pre-allocate for efficiency
+				resp.peers.reserve(resp.peers.size() + static_cast<std::size_t>(len / 18));
+
+				for (int i = 0; i < len; i += 18)
+				{
+					peer_entry p;
+					address_v6::bytes_type addr_bytes;
+					std::memcpy(addr_bytes.data(), peers + i, 16);
+					p.hostname = address_v6(addr_bytes).to_string();
+
+					// read_uint16 expects a mutable reference to advance the pointer
+					// Create a local pointer that won't affect our iteration
+					char const* port_ptr = peers + i + 16;
+					p.port = read_uint16(port_ptr);
+					// Note: port_ptr is now advanced by 2, but we don't use it again
+
+					resp.peers.push_back(std::move(p));
+				}
+			}
+		}
+
+		bdecode_node const ip_ent = e.dict_find_string("external ip");
+		if (ip_ent)
+		{
+			char const* ip_ptr = ip_ent.string_ptr();
+			int const ip_len = ip_ent.string_length();
+
+			if (ip_len == 4)
+			{
+				resp.external_ip = address_v4(read_uint32(ip_ptr));
+			}
+			else if (ip_len == 16)
+			{
+				address_v6::bytes_type addr_bytes;
+				std::memcpy(addr_bytes.data(), ip_ptr, 16);
+				resp.external_ip = address_v6(addr_bytes);
+			}
+		}
+
+		return resp;
+	}
+
+tracker_response parse_scrape_response(
+	span<char const> data,
+	error_code& ec)
+{
+		tracker_response resp;
+
+		bdecode_node e;
+		int const res = bdecode(data.begin(), data.end(), e, ec);
+
+		if (res != 0 || e.type() != bdecode_node::dict_t)
+		{
+			ec = errors::invalid_tracker_response;
+			return resp;
+		}
+
+				if (auto const failure = e.dict_find_string("failure reason"); failure)
+		{
+			resp.failure_reason = std::string(failure.string_value());
+			ec = errors::tracker_failure;
+			return resp;
+		}
+
+		bdecode_node const files = e.dict_find_dict("files");
+		if (!files)
+		{
+			ec = errors::invalid_tracker_response_length;
+			return resp;
+		}
+
+		// We're looking for the first (and should be only) file entry
+		// The key should be the info hash we scraped for
+		if (files.dict_size() > 0)
+		{
+			// Since we don't have the actual info hash here, we'll just get the first entry
+			for (int i = 0; i < files.dict_size(); ++i)
+			{
+				auto [key, value] = files.dict_at(i);
+				if (value.type() == bdecode_node::dict_t)
+				{
+					bdecode_node const scrape_data = value;
+					resp.complete = int(scrape_data.dict_find_int_value("complete", -1));
+					resp.incomplete = int(scrape_data.dict_find_int_value("incomplete", -1));
+					resp.downloaded = int(scrape_data.dict_find_int_value("downloaded", -1));
+					break;
+				}
+			}
+		}
+
+		return resp;
+	}
+
+curl_tracker_client::curl_tracker_client(
+	io_context&,
+	std::string const& url,
+	settings_pack const& settings,
+	std::shared_ptr<curl_thread_manager> curl_mgr)
```

Comment:
passing in ownership of this object here creates a circular reference. I think this should be a raw reference